### PR TITLE
Support additional CodeBuild settings, retry for IAM errors

### PR DESCRIPTION
**Pull request by errriclee**
Support CodeBuild environment variable type, git clone depth, insecure SSL options
Retry UpdateProject similar to CreateProject

### CODE REVIEW

**Comment by bflad** [comment]
Supporting Git clone depth already has a pending PR: #3467 

Can you please work with the original author or implement on top of their commits? The main issue with adding these new attributes inside `source` is that since `source` is a `TypeSet`, it requires a state migration (either to `TypeList` or an updated `TypeSet` hash)

**Comment by kwent** [comment]
Looking forward to this !

**Comment by kwent** [comment]
They also added `branchFilter` https://github.com/aws/aws-sdk-go/commit/c0ea8ae46aeed5902eeebe10d0d5cdf54b258526#diff-4ac32a78649ca5bdd8e0ba38b7006a1eR8

**Comment by errriclee** [comment]
I don't plan on supporting the webhook branch filter in this PR, but it seems separate enough it could be in a separate PR without stepping on toes.

**Comment by errriclee** [comment]
@bflad is this good to go?

File: aws/resource_aws_codebuild_project.go [hunk 1]
```diff
@@ -527,7 +557,22 @@ func resourceAwsCodeBuildProjectUpdate(d *schema.ResourceData, meta interface{})
 	// But its a slice of pointers so if not set for every update, they get removed.
 	params.Tags = tagsFromMapCodeBuild(d.Get("tags").(map[string]interface{}))
 
-	_, err := conn.UpdateProject(params)
+	err := resource.Retry(5*time.Minute, func() *resource.RetryError {
```
[the remark this answers is not in the document]
Comment (errriclee) [reply aws/resource_aws_codebuild_project.go]: Now #4238

**Comment by bflad** [comment]
After looking through the various duplicate pull requests, I'm going to try to get the commits merged in the following order here (resolving the merge conflicts and test changes along the way, which could be tedious): #2811 -> #4021 -> #3929. 🤞 

In the end the following new arguments will be supported:
* `environment` > `environment_variable` > `type`
* `source` > `git_clone_depth`
* `source` > `insecure_ssl`

**Comment by bflad** [comment]
After merging together the above pull requests and refactoring the test suite, we should be good to go:

```
make testacc TEST=./aws TESTARGS='-run=TestAccAWSCodeBuildProject'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSCodeBuildProject -timeout 120m
=== RUN   TestAccAWSCodeBuildProject_basic
--- PASS: TestAccAWSCodeBuildProject_basic (31.87s)
=== RUN   TestAccAWSCodeBuildProject_BadgeEnabled
--- PASS: TestAccAWSCodeBuildProject_BadgeEnabled (30.05s)
=== RUN   TestAccAWSCodeBuildProject_BuildTimeout
--- PASS: TestAccAWSCodeBuildProject_BuildTimeout (38.35s)
=== RUN   TestAccAWSCodeBuildProject_Cache
--- PASS: TestAccAWSCodeBuildProject_Cache (59.77s)
=== RUN   TestAccAWSCodeBuildProject_Description
--- PASS: TestAccAWSCodeBuildProject_Description (38.08s)
=== RUN   TestAccAWSCodeBuildProject_EncryptionKey
--- PASS: TestAccAWSCodeBuildProject_EncryptionKey (53.94s)
=== RUN   TestAccAWSCodeBuildProject_Environment_EnvironmentVariable_Type
--- PASS: TestAccAWSCodeBuildProject_Environment_EnvironmentVariable_Type (30.31s)
=== RUN   TestAccAWSCodeBuildProject_Source_Auth
--- PASS: TestAccAWSCodeBuildProject_Source_Auth (20.33s)
=== RUN   TestAccAWSCodeBuildProject_Source_GitCloneDepth
--- PASS: TestAccAWSCodeBuildProject_Source_GitCloneDepth (38.51s)
=== RUN   TestAccAWSCodeBuildProject_Source_InsecureSSL
--- PASS: TestAccAWSCodeBuildProject_Source_InsecureSSL (37.43s)
=== RUN   TestAccAWSCodeBuildProject_Tags
--- PASS: TestAccAWSCodeBuildProject_Tags (28.91s)
=== RUN   TestAccAWSCodeBuildProject_VpcConfig
--- PASS: TestAccAWSCodeBuildProject_VpcConfig (57.63s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	465.209s
```

This will release with v1.21.0 of the AWS provider, likely middle of next week.

**Comment by bflad** [comment]
This has been released in [version 1.21.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1210-may-31-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**